doc/overviews/inference.md

```diff
@@ -0,0 +1,412 @@
+
+A Kuadrant inference custom resource, such as "PromptGuardPolicy" or "TokenRateLimitPolicy":
+
+1. Targets Gateway API networking resources such as [HTTPRoutes](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPRoute) and [Gateways](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.Gateway), using these resources to obtain additional context, i.e., which traffic workload (HTTP attributes, hostnames, user attributes, etc) to apply inference policies to.
+2. Supports targeting subsets (sections) of a network resource to apply the policies to.
+3. Also supports targeting a specific inference 'model', specified in the request body.
+4. Enables cluster operators to set defaults that govern behavior at the lower levels of the network, until a more specific policy is applied.
+5. Enables platform engineers to set overrides over policies and/or individual policy rules specified at the lower levels of the network.
+
+## How it works
+
+### Envoy's External Processing Protocol
+
+Kuadrant integrates an [External Processing](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_proc_filter) service ("__REPLACE_WITH_NAME_OF_INFERENCE_SERVICE__") that is triggered on matching HTTP contexts.
+
+The workflow per request goes:
+
+1. On incoming request, the gateway checks the matching rules for enforcing the inference policy rules, as stated in the PromptGuardPolicy/TokenRateLimitPolicy custom resources and targeted Gateway API networking objects
+2. If the request matches, the gateway sends a [ProcessingRequest](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/ext_proc/v3/external_processor.proto#envoy-v3-api-msg-service-ext-proc-v3-processingrequest) to __REPLACE_WITH_NAME_OF_INFERENCE_SERVICE__ on the specified grpc port for that service.
+3. The external inference service responds with a [ProcessingResponse](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/ext_proc/v3/external_processor.proto#service-ext-proc-v3-processingresponse) back to the gateway for each stage of the request (e.g. request headers, request body, response headers, response body)
+
+An inference policy custom resource and its targeted Gateway API networking resource contain all the statements to configure both the ingress gateway and the external inference service. The external inference service is configure indirectly via additional context set by the wasm-shim at request time.
+
+### The PromptGuardPolicy & TokenRateLimitPolicy custom resources
+
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: PromptGuardPolicy
+metadata:
+  name: prompt-guard
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: Gateway
+    name: my-llm-gateway
+  model:
+    url: http://huggingface-granite-guardian-default.example.com/v1
+    apiKey:
+      secretRef:
+        name: granite-api-key
+        key: token
+  categories:
+    - harm
+    - violence
+    - sexual_content
```

Comment:
From the point of view of defaults and overrides, especially atomic and merge strategies this configuration needs to be explored more. How would a merge happen?

Reply:
Agreed.
I'm tending towards 'atomic' strategy for a simpler understanding and implementation for now?
Other strategies could bubble up as use cases in the future.